docs/idiomatic-maya-guide.md

```diff
@@ -1,99 +1,179 @@
 
+This document lists some of the guidelines and their corresponding examples that we are trying to follow in Maya to be idiomatic in true sense (there could be difference in opinions but we are trying to make sure that we all remain on the same page and hence it doesn't look confusing going further) -
 
-## Devil is in the details
-Below are some sample code reviews that describes with examples to write idiomatic maya code.
+## Naming Conventions
 
-### Naming - Thumb Rule
-- Use names based on what the logic provides and not based on what the logic contains
+- Try to use names based on what the logic provides and not based on what the logic contains
 
 ```go
 import (
+  // Bad
   // alias `algorithm` conveys about what it contains i.e. some algorithm
   algorithm "github.com/openebs/maya/pkg/algorithm/nodeSelect/v1alpha1"
 
   // vs.
 
+  // Good
   // `nodeselect` tries to convey what the logic provides
   // this seems more natural way to express
   nodeselect "github.com/openebs/maya/pkg/algorithm/nodeSelect/v1alpha1"
 )
 ```
 
 ```go
-nodeDisks, err := ac.NodeDiskSelector()
-// `disks` is repeated; does not seem natural
-if len(nodeDisks.Disks.Items) == 0 {
-  // ...
-}
+import(
+// Bad
+// upgrade looks more generic i.e upgrade package can have
+// different other sub-packages like result which can be imported
+upgrade "github.com/openebs/maya/pkg/upgrade/result/v1alpha1"
+
+// vs
+
+// Good
+// upgraderesult looks more specific and to the point i.e. the alias
+// is for 'result' which is inside upgrade
+upgraderesult "github.com/openebs/maya/pkg/upgrade/result/v1alpha1"
+```
+
+```go
+// Bad
+// taskPatch represents details for runtask patch operation
+type taskPatch struct {}
+
+// vs
+
+// Good
+// runTaskPatch represents details for runtask patch operation
+type runTaskPatch struct {}
+```
+
+```go
+// redundant (patch) -- not ok
+patch.IsValidPatchType()
 
 // vs.
 
-s, err := ac.NodeDiskSelector()
-if len(s.Disks.Items) == 0 {
-  // ...
-}
+// Good
+patch.IsValidType()
+```
+
+## Indentation (For enhancing readability)
+
+We try to keep a line of code limited to not more than 80 characters so that the visibility and readability of the code looks better and aligned.
+
+```go
+// Bad
+p, err := patch.BuilderForRuntask("UpgradeResult", m.runtask.Spec.Task, m.templateValues).AddCheck(patch.IsValidType()).Build()
+
+// vs
+
+// Good (looks more readable if aligned this way)
+p, err := patch.
+  BuilderForRuntask("UpgradeResult", m.runtask.Spec.Task, m.templateValues).
+  AddCheck(patch.IsValidType()).
+  Build()
 ```
 
-### Do the names reflect their purpose?
 ```go
-// assume a file pkg/algorithm/nodeSelect/v1alpha1/select_node.go 
+// Bad
```

Comment:
We might want to have separate go based snippet for bad vs. good code.
e.g.
```go
// bad code
```
```go
// good code
```

Notice two code snippets & not a single go snippet.

Reply:
Done